File: lab1/src/BubbleSort.java
import java.util.Comparator;
import java.util.List;

public class BubbleSort {

    public static <T> void sort(List<T> list, Comparator<T> comparator){
        int size = list.size();
        for (int i = 1; i < size; ++i) {
            for (int left = 0; left < (size - i); ++left) {
                int right = left + 1;
                if (comparator.compare(list.get(left), list.get(right)) > 0)
                    swap(list, left, right);
            }
        }
    }

    private static <T> void swap(List<T> list, int left, int right){
        T temp = list.get(left);
        list.set(left, list.get(right));
        list.set(right, temp);
    }
}

File: BubbleSort.scala

object BubbleSort {

  def bubbleSort(array: Array[Double]): Array[Double] = {
    var swap = false

    for (i <- 0 until array.length - 1) {
      if (array(i + 1) < array(i)) {
        val temp = array(i)
        array(i) = array(i + 1)
        array(i + 1) = temp
        swap = true
      }
    }
    if(swap)
      bubbleSort(array)
      else
        array
  }

  def main(args: Array[String]): Unit = {
      var myList = Array(23.5, -54.63, 67.4, -533.544, 0, 563.41)
      bubbleSort(myList);
  }
}

